client2/thin.go

```diff
@@ -0,0 +1,308 @@
+// SPDX-FileCopyrightText: © 2023 David Stainton
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package client2
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"sync"
+
+	"github.com/charmbracelet/log"
+	"github.com/fxamacker/cbor/v2"
+
+	cpki "github.com/katzenpost/katzenpost/core/pki"
+	sConstants "github.com/katzenpost/katzenpost/core/sphinx/constants"
+	"github.com/katzenpost/katzenpost/core/worker"
+)
+
+// ThinResponse is used to encapsulate a message response
+// that are passed to the client application.
+type ThinResponse struct {
+
+	// SURBID, a unique indentifier for this response,
+	// which should precisely match the application's chosen
+	// SURBID of the sent message.
+	SURBID *[sConstants.SURBIDLength]byte
+
+	ID *[MessageIDLength]byte
+
+	// Payload is the decrypted payload plaintext.
+	Payload []byte
+}
+
+// ThinClient is the client that handles communication between the mixnet application
+// and the client daemon. It does not do any encryption or decryption or checking
+// of cryptographic signatures; those responsibilities are left to the client daemon
+// process.
+type ThinClient struct {
+	worker.Worker
+
+	log          *log.Logger
+	unixConn     *net.UnixConn
+	destUnixAddr *net.UnixAddr
+
+	pkidoc      *cpki.Document
+	pkidocMutex sync.RWMutex
+
+	isConnected bool
+
+	receivedCh chan ThinResponse
+}
+
+// NewThinClient creates a new thing client.
+func NewThinClient() *ThinClient {
+	return &ThinClient{
+		log: log.NewWithOptions(os.Stderr, log.Options{
+			Prefix: "thin_client",
+			Level:  log.DebugLevel,
+		}),
+		receivedCh: make(chan ThinResponse),
+	}
+}
+
+// Close halts the thin client worker thread and closes the socket
+// connection with the client daemon.
+func (t *ThinClient) Close() error {
+	err := t.unixConn.Close()
+	t.Halt()
+	return err
+}
+
+// Dial dials the client daemon via our agreed upon abstract unix domain socket.
+func (t *ThinClient) Dial() error {
+	t.log.Debug("Dial begin")
+	srcUnixAddr, err := net.ResolveUnixAddr("unixpacket", "@katzenpost_golang_thin_client")
+	if err != nil {
+		return err
+	}
+
+	t.destUnixAddr, err = net.ResolveUnixAddr("unixpacket", "@katzenpost")
+	if err != nil {
+		return err
+	}
+
+	t.log.Debugf("Dial unixpacket %s %s", srcUnixAddr, t.destUnixAddr)
+	t.unixConn, err = net.DialUnix("unixpacket", srcUnixAddr, t.destUnixAddr)
+	if err != nil {
+		return err
+	}
+
+	// WAIT UNTIL we have a Noise cryptographic connection with an edge node
+	t.log.Debugf("Waiting for a connection status message")
+	message1, err := t.readNextMessage()
+	if err != nil {
+		return err
+	}
+	if !message1.IsStatus {
+		panic("did not receive a connection status message")
```

Comment:
```suggestion
		panic("did not receive a connection status message as the first message on fresh thin client connection")
```
Or something like that; the important part is that we expect `IsStatus` as the very first thing. We should probably log what the hell else we got instead.